FiftyOne.Caching.Tests/Loaders/TrackingLoaderBase.cs

```diff
@@ -53,44 +64,90 @@ public TrackingLoaderBase() : this(0)
 
         public TrackingLoaderBase(int delayMillis) : this(delayMillis, false)
         {
+        }
+
+        /// <param name="delayMillis">
+        /// Delay before result is returned in ms.
+        /// </param>
+        /// <param name="runWithToken">
+        /// Whether the canellation of the token
+        /// should cancel a task (true)
+        /// or the internal loop operation (false).
+        /// </param>
+        public TrackingLoaderBase(int delayMillis, bool runWithToken) :
+            this(
+                delayMillis,
+                runWithToken ? null : static t => new CancellationToken(),
+                runWithToken ? static t => new CancellationToken() : null)
+        { 
         }
-
-        public TrackingLoaderBase(int delayMillis, bool runWithToken)
+
+        /// <param name="delayMillis">
+        /// Delay before result is returned in ms.
+        /// </param>
+        /// <param name="tokenForTask">
+        /// Transforms a token to be passed for task cancellation.
+        /// If is `null` or returns `null`, the token from `Load` is used.
+        /// </param>
+        /// <param name="tokenForLoading">
+        /// Transforms a token to be passed for a loop to track.
+        /// If is `null` or returns `null`, the token from `Load` is used.
+        /// </param>
+        public TrackingLoaderBase(
+            int delayMillis, 
+            Func<CancellationToken, CancellationToken> tokenForTask,
+            Func<CancellationToken, CancellationToken> tokenForLoading)
         {
             _delayMillis = delayMillis;
-            _runWithToken = runWithToken;
+            _tokenForTask = tokenForTask;
+            _tokenForLoading = tokenForLoading;
         }
 
         public Task<TValue> Load(TKey key, CancellationToken token)
         {
-            Interlocked.Increment(ref _calls);
+            lock (_countersLock) {
+                ++_calls;
+            }
+            var tokenForTask = _tokenForTask?.Invoke(token) ?? token;
+            var tokenForLoading = _tokenForLoading?.Invoke(token) ?? token;
```

Comment:
these manipulations are super complicated... in the constructor we pass these closures that simply override the passed `token`..  do we use it only for tests?  in that case maybe there is some simpler way to test this, mocking some objects etc?